xbout/plotting/plotfuncs.py

```diff
@@ -0,0 +1,170 @@
+                                     infer_intervals=False,
+                                     add_colorbar=False, ax=ax, **kwargs)]
+    if rest:
+        for region in rest:
+            artist = region.plot.pcolormesh(x=x, y=y, vmin=0, vmax=1,
+                                            cmap='tab20',
+                                            infer_intervals=False,
+                                            add_colorbar=False,
+                                            add_labels=False, ax=ax, **kwargs)
+            artists.append(artist)
+    return artists
+
+
+def plot2d_wrapper(da, method, *, ax=None, separatrix=True, targets=True,
+                   add_limiter_hatching=True, cmap=None, vmin=None, vmax=None,
+                   aspect=None, **kwargs):
+    """
+    Make a 2D plot using an xarray method, taking into account branch cuts (X-points).
+
+    Wraps `xarray.plot` methods, so automatically adds labels.
+
+    Parameters
+    ----------
+    da : xarray.DataArray
+        A 2D (x,y) DataArray of data to plot
+    method : xarray.plot.*
+        An xarray plotting method to use
+    ax : Axes, optional
+        A matplotlib axes instance to plot to. If None, create a new
+        figure and axes, and plot to that
+    separatrix : bool, optional
+        Add dashed lines showing separatrices
+    targets : bool, optional
+        Draw solid lines at the target surfaces
+    add_limiter_hatching : bool, optional
+        Draw hatched areas at the targets
+    cmap : Matplotlib colormap, optional
+        Color map to use for the plot
+    vmin : float, optional
+        Minimum value for the color scale
+    vmax : float, optional
+        Maximum value for the color scale
+    aspect : str or float, optional
+        Passed to ax.set_aspect(). By default 'equal' is used.
+    levels : int or iterable, optional
+        Only used by contour or contourf, sets the number of levels (if int) or the level
+        values (if iterable)
+    **kwargs : optional
+        Additional arguments are passed on to method
+
+    Returns
+    -------
+    artists
+        List of the artist instances
+    """
+
+    # TODO generalise this
+    x = kwargs.pop('x', 'R')
+    y = kwargs.pop('y', 'Z')
+
+    if len(da.dims) != 2:
+        raise ValueError("da must be 2D (x,y)")
+
+    if ax is None:
+        fig, ax = plt.subplots()
+    else:
+        fig = ax.get_figure()
+
+    if aspect is None:
+        aspect = 'equal'
+    ax.set_aspect(aspect)
+
+    if vmin is None:
+        vmin = da.min().values
+    if vmax is None:
+        vmax = da.max().values
+
+    if method is xr.plot.contour or method is xr.plot.contourf:
+        levels = kwargs.get('levels', 7)
+        if isinstance(levels, np.int):
+            levels = np.linspace(vmin, vmax, levels, endpoint=True)
+            # put levels back into kwargs
+            kwargs['levels'] = levels
+        else:
+            levels = np.array(levels)
```

Comment:
Why will it fail? If there's no `'levels'` in `kwargs` then the default `7` will go into the first case, which sets `kwargs['levels']`, to get into the second case `kwargs['levels']` must have been set. This should be `np.array(list(levels))` in case an iterator is passed though.

Reply:
Sorry, I see now, I need `levels` to be a numpy array further down!